File: ShanYin/ERP/CS/Client/src/api/endpoints/logistics.ts
import { apiClient } from '../client'

// Backend constants (Chinese)
export type LogisticsStatus = '待发货' | '在途' | '签收' | '完成' | '取消'
export type ExpressStatus = '待发货' | '在途' | '签收'

export interface ExpressItem {
  sku_id: number
  sku_name: string
  qty: number
}

// 服务端实际返回的地址信息（中文 key）
export interface AddressInfo {
  收货点位Id: number
  收货点位名称: string
  收货地址: string
  收货联系电话: string | null
  发货点位Id: number
  发货点位名称: string
  发货地址: string
  发货联系电话: string | null
}

export interface ExpressOrder {
  id: number
  tracking_number: string
  status: string
  address_info: AddressInfo
  items: ExpressItem[]
}

export interface Logistics {
  id: number
  virtual_contract_id: number
  status: string
  created_at: string
  express_orders_count: number
}

export interface LogisticsDetail extends Logistics {
  express_orders: ExpressOrder[]
  vc_type?: string
  transaction_date?: string
  elements?: Array<{
    shipping_point_id?: number
    shipping_point_name?: string
    receiving_point_id?: number
    receiving_point_name?: string
    sku_id: number
    sku_name: string
    qty: number
    price?: number
  }>
}

export interface LogisticsListResponse {
  items: Logistics[]
  total: number
  page: number
  size: number
}

export interface LogisticsDashboardSummary {
  logistics_summary: {
    total: number
    pending: number
    transit: number
    signed: number
    finish: number
    today_new: number
  }
  express_summary: {
    total: number
    pending: number
    transit: number
    signed: number
  }
}

export interface CreateLogisticsPlanSchema {
  vc_id: number
  orders: {
    tracking_number: string
    items: { sku_id: number; qty: number }[]
    address_info: AddressInfo
  }[]
  created_date?: string
}

export interface BatchItem {
  sku_id: number
  production_date: string
  receiving_point_id: number
  qty: number
  certificate_filename?: string
}

export interface ConfirmInboundSchema {
  log_id: number
  sn_list: string[]
  batch_items?: BatchItem[]
  created_date?: string
}

export interface ExpressOrderStatusSchema {
  order_id: number
  target_status: ExpressStatus
  logistics_id: number
  created_date?: string
}

export interface BatchItemWithFile extends BatchItem {
  certificate_file?: File
}

// Express Order Global Overview types
export interface ExpressOrderGlobalItem {
  id: number
  tracking_number: string
  status: ExpressStatus
  created_at: string
  transaction_date?: string
  logistics_id: number
  items: Array<{ sku_id: number; sku_name: string; qty: number }>
  address_info: AddressInfo
  vc_id: number | null
  vc_type: string | null
  vc_subject_status: string | null
}

export interface ExpressOrderGlobalResponse {
  items: ExpressOrderGlobalItem[]
  total: number
  page: number
  size: number
}

export interface ExpressOrderGlobalParams {
  ids?: number
  tracking_number?: string
  status?: ExpressStatus
  date_from?: string
  date_to?: string
  sku_id?: number
  sku_name_kw?: string
  shipping_point_id?: number
  shipping_point_name_kw?: string
  receiving_point_id?: number
  receiving_point_name_kw?: string
  vc_id?: number
  vc_type?: string
  vc_status_type?: '主状态' | '合同状态'
  vc_status_value?: string
  subject_status?: string
  business_id?: number
  business_customer_name_kw?: string
  supply_chain_id?: number
  supply_chain_supplier_name_kw?: string
  page?: number
  size?: number
}

// Logistics Global Overview types
export interface LogisticsGlobalItem {
  id: number
  virtual_contract_id: number
  status: LogisticsStatus
  created_at: string
  transaction_date?: string
  express_orders_count: number
  vc_type: string | null
}

export interface LogisticsGlobalResponse {
  items: LogisticsGlobalItem[]
  total: number
  page: number
  size: number
}

export interface LogisticsGlobalParams {
  ids?: number
  status?: LogisticsStatus
  date_from?: string
  date_to?: string
  tracking_number?: string
  express_order_id?: number
  sku_id?: number
  sku_name_kw?: string
  shipping_point_id?: number
  shipping_point_name_kw?: string
  receiving_point_id?: number
  receiving_point_name_kw?: string
  vc_id?: number
  vc_type?: string
  vc_status_type?: '主状态' | '合同状态'
  vc_status_value?: string
  subject_status?: string
  business_id?: number
  business_customer_name_kw?: string
  supply_chain_id?: number
  supply_chain_supplier_name_kw?: string
  page?: number
  size?: number
}

export const logisticsApi = {
  list: (params?: {
    ids?: number[]
    vc_id?: number
    status?: string
    date_from?: string
    date_to?: string
    tracking_number?: string
    page?: number
    size?: number
  }) => apiClient.get<LogisticsListResponse>('/logistics/list', { params }) as unknown as Promise<LogisticsListResponse>,

  getDetail: (logId: number) =>
    apiClient.get<LogisticsDetail>(`/logistics/${logId}`) as unknown as Promise<LogisticsDetail>,

  createPlan: (data: CreateLogisticsPlanSchema) =>
    apiClient.post<{ success: boolean }>('/logistics/create-plan', data) as unknown as Promise<{ success: boolean }>,

  confirmInbound: (data: ConfirmInboundSchema) =>
    apiClient.post<{ success: boolean }>('/logistics/confirm-inbound', data) as unknown as Promise<{ success: boolean }>,

  confirmInboundMaterial: (data: {
    log_id: number
    sn_list: string[]
    batch_items: BatchItem[]
    certificates: File[]
    created_date?: string
  }) => {
    const formData = new FormData()
    formData.append('log_id', String(data.log_id))
    formData.append('sn_list', JSON.stringify(data.sn_list))
    formData.append('batch_items_json', JSON.stringify(data.batch_items))
    if (data.created_date) formData.append('created_date', data.created_date)
    data.certificates.forEach((file) => formData.append('certificates', file))
    return apiClient.post('/logistics/confirm-inbound-material', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }) as unknown as Promise<{ success: boolean }>
  },

  updateExpressOrder: (data: {
    order_id: number
    tracking_number: string
    address_info: AddressInfo
  }) => apiClient.put<{ success: boolean }>('/logistics/update-express', data) as unknown as Promise<{ success: boolean }>,

  updateExpressStatus: (data: ExpressOrderStatusSchema) =>
    apiClient.post<{ success: boolean }>('/logistics/update-express-status', data) as unknown as Promise<{ success: boolean }>,

  bulkProgress: (data: {
    order_ids: number[]
    target_status: ExpressStatus
    logistics_id: number
    created_date?: string
  }) => apiClient.post<{ success: boolean }>('/logistics/bulk-progress', data) as unknown as Promise<{ success: boolean }>,

  getDashboardSummary: () =>
    apiClient.get<LogisticsDashboardSummary>('/logistics/dashboard/summary') as unknown as Promise<LogisticsDashboardSummary>,

  getExpressOrdersGlobal: (params: ExpressOrderGlobalParams) =>
    apiClient.get<ExpressOrderGlobalResponse>('/logistics/express-orders/global', { params }) as unknown as Promise<ExpressOrderGlobalResponse>,

  getLogisticsGlobal: (params: LogisticsGlobalParams) =>
    apiClient.get<LogisticsGlobalResponse>('/logistics/global', { params }) as unknown as Promise<LogisticsGlobalResponse>,
}

File: ShanYin/ERP/CS/Server/api/routers/logistics.py
from fastapi import APIRouter, Depends, UploadFile, File, Form
from pydantic import BaseModel
from typing import List, Optional
from datetime import date
from sqlalchemy.orm import Session
import json
import os
from api.deps import get_db, verify_token, api_success, parse_ids
from api.middleware.error_handler import raise_not_found_error
from logic.api_queries import list_logistics, get_logistics, list_express_orders_global, list_logistics_global
from logic.logistics import (
    create_logistics_plan_action, confirm_inbound_action,
    update_express_order_action, update_express_order_status_action,
    bulk_progress_express_orders_action,
    CreateLogisticsPlanSchema, ConfirmInboundSchema,
    UpdateExpressOrderSchema, ExpressOrderStatusSchema,
    BatchItemSchema,
)
from logic.file_mgmt import save_batch_certificate
from logic.logistics.queries import get_logistics_dashboard_summary

router = APIRouter(prefix="/api/v1/logistics", tags=["物流"], dependencies=[Depends(verify_token)])


@router.post("/create-plan", summary="创建物流发货计划")
def create_logistics_plan(payload: CreateLogisticsPlanSchema, session: Session = Depends(get_db)):
    return create_logistics_plan_action(session, payload).model_dump()


@router.post("/confirm-inbound", summary="确认入库")
def confirm_inbound(payload: ConfirmInboundSchema, session: Session = Depends(get_db)):
    return confirm_inbound_action(session, payload).model_dump()


@router.post("/confirm-inbound-material", summary="物料采购确认入库（含质检报告）")
async def confirm_inbound_material(
    log_id: int = Form(...),
    sn_list: str = Form("[]"),
    batch_items_json: str = Form(..., description="批次明细 JSON 字符串"),
    certificates: List[UploadFile] = File(default=[]),
    created_date: Optional[str] = Form(None, description="业务发生日期 YYYY-MM-DD"),
    session: Session = Depends(get_db)
):
    sn_parsed = json.loads(sn_list)
    batch_items_raw = json.loads(batch_items_json)
    batch_items = [BatchItemSchema(**bi) for bi in batch_items_raw]

    # 构建 certificate_filename → batch_item 索引 的映射（client 传 batch_no 作为 certificate_filename）
    cert_key_to_idx: dict[str, int] = {}
    for idx, bi in enumerate(batch_items):
        if bi.certificate_filename:
            cert_key_to_idx[bi.certificate_filename] = idx

    # 按 certificate_filename 分组保存证书文件
    # cert_paths: batch_item_idx → [saved_path, ...]
    cert_paths: dict[int, list[str]] = {i: [] for i in range(len(batch_items))}
    for cert_file in certificates:
        if not cert_file.filename:
            continue
        fname = os.path.splitext(cert_file.filename)[0]  # client 传的 batch_no
        ext = os.path.splitext(cert_file.filename)[1]
        idx = cert_key_to_idx.get(fname)
        if idx is None:
            continue
        batch_no = fname  # batch_no 就是 certificate_filename
        existing_count = len(cert_paths[idx])
        # 命名：单文件={batch_no}{ext}，多文件={batch_no}_1.ext, {batch_no}_2.ext, ...
        cert_filename = f"{batch_no}_{existing_count + 1}{ext}" if existing_count > 0 else f"{batch_no}{ext}"
        saved_path = save_batch_certificate(cert_filename, cert_file)
        cert_paths[idx].append(saved_path)

    # 将证书路径写回 batch_items
    for idx, bi in enumerate(batch_items):
        paths = cert_paths[idx]
        if paths:
            bi.certificate_filename = json.dumps(paths) if len(paths) > 1 else paths[0]

    from datetime import datetime as dt
    created_date_parsed = dt.strptime(created_date, "%Y-%m-%d").date() if created_date else None
    payload = ConfirmInboundSchema(log_id=log_id, sn_list=sn_parsed, batch_items=batch_items, created_date=created_date_parsed)
    return confirm_inbound_action(session, payload).model_dump()


@router.put("/update-express", summary="更新快递单信息")
def update_express_order(payload: UpdateExpressOrderSchema, session: Session = Depends(get_db)):
    return update_express_order_action(session, payload).model_dump()


@router.post("/update-express-status", summary="更新快递状态")
def update_express_order_status(payload: ExpressOrderStatusSchema, session: Session = Depends(get_db)):
    return update_express_order_status_action(session, payload).model_dump()


class BulkProgressRequest(BaseModel):
    order_ids: List[int]
    target_status: str
    logistics_id: int
    created_date: Optional[date] = None

@router.post("/bulk-progress", summary="批量推进快递状态")
def bulk_progress_express_orders(req: BulkProgressRequest, session: Session = Depends(get_db)):
    return bulk_progress_express_orders_action(
        session, req.order_ids, req.target_status, req.logistics_id, created_date=req.created_date).model_dump()


# ==================== Query Endpoints ====================

@router.get("/list", summary="物流列表")
def get_logistics_list(
    ids: Optional[str] = None,
    vc_id: Optional[int] = None,
    status: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    tracking_number: Optional[str] = None,
    page: int = 1,
    size: int = 50,
    session: Session = Depends(get_db)
):
    id_list = parse_ids(ids)
    result = list_logistics(session, ids=id_list, vc_id=vc_id, status=status,
                             date_from=date_from, date_to=date_to,
                             tracking_number=tracking_number, page=page, size=size)
    return api_success(result)


@router.get("/express-orders/global", summary="快递单全局概览")
def get_express_orders_global(
    ids: Optional[int] = None,
    tracking_number: Optional[str] = None,
    status: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    sku_id: Optional[int] = None,
    sku_name_kw: Optional[str] = None,
    shipping_point_id: Optional[int] = None,
    shipping_point_name_kw: Optional[str] = None,
    receiving_point_id: Optional[int] = None,
    receiving_point_name_kw: Optional[str] = None,
    vc_id: Optional[int] = None,
    vc_type: Optional[str] = None,
    vc_status_type: Optional[str] = None,
    vc_status_value: Optional[str] = None,
    subject_status: Optional[str] = None,
    business_id: Optional[int] = None,
    business_customer_name_kw: Optional[str] = None,
    supply_chain_id: Optional[int] = None,
    supply_chain_supplier_name_kw: Optional[str] = None,
    page: int = 1,
    size: int = 20,
    session: Session = Depends(get_db)
):
    id_list = [ids] if ids else None
    result = list_express_orders_global(
        session, ids=id_list, tracking_number=tracking_number, status=status,
        date_from=date_from, date_to=date_to,
        sku_id=sku_id, sku_name_kw=sku_name_kw,
        shipping_point_id=shipping_point_id, shipping_point_name_kw=shipping_point_name_kw,
        receiving_point_id=receiving_point_id, receiving_point_name_kw=receiving_point_name_kw,
        vc_id=vc_id, vc_type=vc_type, vc_status_type=vc_status_type, vc_status_value=vc_status_value,
        subject_status=subject_status,
        business_id=business_id, business_customer_name_kw=business_customer_name_kw,
        supply_chain_id=supply_chain_id, supply_chain_supplier_name_kw=supply_chain_supplier_name_kw,
        page=page, size=size
    )
    return api_success(result)


@router.get("/global", summary="物流全局概览")
def get_logistics_global(
    ids: Optional[int] = None,
    status: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    tracking_number: Optional[str] = None,
    express_order_id: Optional[int] = None,
    sku_id: Optional[int] = None,
    sku_name_kw: Optional[str] = None,
    shipping_point_id: Optional[int] = None,
    shipping_point_name_kw: Optional[str] = None,
    receiving_point_id: Optional[int] = None,
    receiving_point_name_kw: Optional[str] = None,
    vc_id: Optional[int] = None,
    vc_type: Optional[str] = None,
    vc_status_type: Optional[str] = None,
    vc_status_value: Optional[str] = None,
    subject_status: Optional[str] = None,
    business_id: Optional[int] = None,
    business_customer_name_kw: Optional[str] = None,
    supply_chain_id: Optional[int] = None,
    supply_chain_supplier_name_kw: Optional[str] = None,
    page: int = 1,
    size: int = 20,
    session: Session = Depends(get_db)
):
    id_list = [ids] if ids else None
    result = list_logistics_global(
        session, ids=id_list, status=status,
        date_from=date_from, date_to=date_to,
        tracking_number=tracking_number, express_order_id=express_order_id,
        sku_id=sku_id, sku_name_kw=sku_name_kw,
        shipping_point_id=shipping_point_id, shipping_point_name_kw=shipping_point_name_kw,
        receiving_point_id=receiving_point_id, receiving_point_name_kw=receiving_point_name_kw,
        vc_id=vc_id, vc_type=vc_type, vc_status_type=vc_status_type, vc_status_value=vc_status_value,
        subject_status=subject_status,
        business_id=business_id, business_customer_name_kw=business_customer_name_kw,
        supply_chain_id=supply_chain_id, supply_chain_supplier_name_kw=supply_chain_supplier_name_kw,
        page=page, size=size
    )
    return api_success(result)


@router.get("/dashboard/summary", summary="物流看板统计")
def get_logistics_dash(session: Session = Depends(get_db)):
    """物流看板：各状态数量统计、今日新增。"""
    return api_success(get_logistics_dashboard_summary())


@router.get("/{log_id}", summary="物流详情")
def get_logistics_detail(log_id: int, session: Session = Depends(get_db)):
    data = get_logistics(session, log_id)
    if data is None:
        raise_not_found_error("物流", str(log_id))
    return api_success(data)
